src/clientiface.cpp

```diff
@@ -46,396 +45,245 @@ const char *ClientInterface::statenames[] = {
+		const WantedMapSend &wms = m_map_send_queue[i];
+		if (wms.type == WMST_MAPBLOCK) {
+			/*verbosestream << "Server: Client "<<peer_id<<" wants MapBlock ("
+					<<wms.p.X<<","<<wms.p.Y<<","<<wms.p.Z<<")"
+					<< std::endl;*/
+
+			// Do not go over-limit
+			if (blockpos_over_max_limit(wms.p))
 				continue;
 
-			/*
-				Do not go over max mapgen limit
-			*/
-			if (blockpos_over_max_limit(p))
+			// Don't send blocks that are currently being transferred
+			if (m_blocks_sending.find(wms) != m_blocks_sending.end())
 				continue;
 
-			// If this is true, inexistent block will be made from scratch
-			bool generate = d <= d_max_gen;
-
-			/*
-				Don't generate or send if not in sight
-				FIXME This only works if the client uses a small enough
-				FOV setting. The default of 72 degrees is fine.
-				Also retrieve a smaller view cone in the direction of the player's
-				movement.
-				(0.1 is about 4 degrees)
-			*/
-			f32 dist;
-			if (!(isBlockInSight(p, camera_pos, camera_dir, camera_fov,
-						d_blocks_in_sight, &dist) ||
-					(playerspeed.getLength() > 1.0f * BS &&
-					isBlockInSight(p, camera_pos, playerspeeddir, 0.1f,
-						d_blocks_in_sight)))) {
-				continue;
+			// Don't send blocks that have already been sent, unless it has been
+			// updated
+			if (m_mapblocks_sent.find(wms.p) != m_mapblocks_sent.end()) {
+				if (m_blocks_updated_since_last_send.count(wms) == 0)
+					continue;
 			}
 
-			/*
-				Don't send already sent blocks
-			*/
-			if (m_blocks_sent.find(p) != m_blocks_sent.end())
-				continue;
+			// If the MapBlock is not loaded, it will be queued to be loaded or
+			// generated. Otherwise it will be added to 'dest'.
 
-			/*
-				Check if map has this block
-			*/
-			MapBlock *block = env->getMap().getBlockNoCreateNoEx(p);
+			const bool allow_generate = true;
 
-			bool block_not_found = false;
-			if (block) {
+			MapBlock *block = m_env->getMap().getBlockNoCreateNoEx(wms.p);
+
+			bool surely_not_found_on_disk = false;
+			bool emerge_required = false;
+			if(block != NULL)
+			{
 				// Reset usage timer, this block will be of use in the future.
 				block->resetUsageTimer();
 
-				// Check whether the block exists (with data)
-				if (block->isDummy() || !block->isGenerated())
-					block_not_found = true;
+				// Block is dummy if data doesn't exist.
+				// It means it has been not found from disk and not generated
+				if(block->isDummy())
+					surely_not_found_on_disk = true;
 
-				/*
-					If block is not close, don't send it unless it is near
-					ground level.
-
-					Block is near ground level if night-time mesh
-					differs from day-time mesh.
-				*/
-				if (d >= d_opt) {
-					if (!block->getIsUnderground() && !block->getDayNightDiff())
-						continue;
-				}
+				if(block->isGenerated() == false && allow_generate)
+					emerge_required = true;
 
-				if (m_occ_cull && !block_not_found &&
-						env->getMap().isBlockOccluded(block, cam_pos_nodes)) {
```

Comment:
Do we rely on the client doing occlusion culling and only requested needed blocks?